### PR TITLE
JavaScript pipeline for new data-capture scripts

### LINKED ISSUE #389: New JavaScript pipeline
Meta-issue for discussing new JS pipeline/build/style/etc.

My preference would be to use ES6 with npm-installed modules, built nicely into browser-ready ES5 via Babel and Browserify (as gulp tasks). However, that might be overkill, and we could potentially just go with what we're already doing, which is ES5, vendor'd libraries, and simple concatenation (via gulp).

Thoughts, @toolness @hbillings?

### PR DESCRIPTION
**Pull request by jseppi**
Ready for review.

- add new `gulp js:data-capture` task for babel+browserify on new data-capture scripts (with sourcemaps)
- add new `gulp clean` task for clearing out all built files
- use `airbnb-base` as the eslint config instead of `airbnb` (which is for react projects)
- various minor gulpfile rejiggerings/renamings

Closes #389

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -1,28 +1,51 @@
...
+      dataCapture: 'hourglass_site/static/hourglass_site/js/built/data-capture',
     },
   },
 };
 
 const paths = {
   sass: '**/*.scss',
   js: '**/*.js',
+  dataCaptureEntry: 'data-capture/index.js',
+  dataCaptureOutry: 'index.min.js',
```
Comment (toolness) [inline gulpfile.js]: I read this as `dataCaptureOutcry` :)

Comment (jseppi) [inline gulpfile.js]: I suppose `dataCaptureOutputFile` would be a better name, since `outry` isn't a real word.

**Comment by toolness** [comment]
Cool! Is `data-capture/index.js` currently being loaded by any pages? I don't really expect it to be, am just curious...

File: gulpfile.js (hunk 1 above)
Comment (toolness) [inline gulpfile.js]: Haha I was wondering about that--figured it was some tech word I hadn't acclimated myself to yet...

**Comment by jseppi** [comment]
Nope. I did a manual check locally of removing `common/upload.js` from the bundle, requiring it in `data-capture/index.js`, and including `built/data-capture/index.js` in the styleguide template to make sure it worked though.

**Comment by toolness** [comment]
Cool!

Ok, so just to see what would happen I added the following near the bottom of `data-explorer/index.js`:

```js
var boop = () => { return "foo"; };
console.log(boop());
```

And it worked--when I visited CALC's homepage, `foo` was logged to my console.

However, I then did some snooping and noticed that `index.min.js`, despite its name, is not actually minified--is this because `DEBUG` is `True`?  Was this the previous behavior too and I just wasn't aware of it?

The other thing I noticed was that `index.min.js` didn't actually transpile my ES2015 code into ES5... It literally appeared in the file just as I wrote it in code, which concerns me, because I'm worried about folks in the future adding ES2015 code to the legacy codebase, assuming it's being transpiled because everything works in their ES2015-compliant browser, and then things breaking for users who are on older browsers.

I then tried adding the same code to the bottom of `data-capture/index.js`, and it actually crashed gulp--it didn't just log a traceback and then continue gulp's watching, it actually made the gulp process exit with a nonzero status code. This was particularly confusing because I don't think my code had any syntax errors...

<details>
  <summary>Gulp Traceback</summary>

```
gulp_1  | [17:34:55] Using gulpfile /calc/gulpfile.js
gulp_1  | [17:34:55] Starting 'sass'...
gulp_1  | [17:34:55] Starting 'lint'...
gulp_1  | [17:34:55] Starting 'js:data-capture'...
gulp_1  | [17:34:55] Finished 'js:data-capture' after 11 ms
gulp_1  | [17:34:55] Starting 'js:data-explorer:index'...
gulp_1  | [17:34:55] Starting 'js:common:base'...
gulp_1  | 
gulp_1  | events.js:154
gulp_1  |       throw er; // Unhandled 'error' event
gulp_1  |       ^
gulp_1  | Error
gulp_1  |     at new JS_Parse_Error (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:1534:18)
gulp_1  |     at js_error (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:1542:11)
gulp_1  |     at croak (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2089:9)
gulp_1  |     at token_error (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2097:9)
gulp_1  |     at unexpected (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2103:9)
gulp_1  |     at expr_atom (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2618:13)
gulp_1  |     at maybe_unary (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2792:19)
gulp_1  |     at expr_ops (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2827:24)
gulp_1  |     at maybe_conditional (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2832:20)
gulp_1  |     at maybe_assign (eval at <anonymous> (/node/node_modules/uglify-js/tools/node.js:28:1), <anonymous>:2856:20)
calc_gulp_1 exited with code 1
```

</details>

**Comment by jseppi** [comment]
Looking into it...

**Comment by toolness** [comment]
Nice! Now the problem seems to be when I *actually* add a syntax error...

<details>
<summary>`TypeError: Cannot read property 'emit' of undefined` w/ nonzero exit from gulp</summary>
```
gulp_1  | [18:05:41] Finished 'lint' after 447 ms
gulp_1  | [18:05:41] Starting 'js'...
gulp_1  | [18:05:41] Finished 'js' after 13 μs
gulp_1  | [18:05:49] Starting 'lint'...
gulp_1  | [18:05:49] Starting 'js:data-capture'...
gulp_1  | [18:05:49] Finished 'js:data-capture' after 4.99 ms
gulp_1  | [18:05:49] Starting 'js:data-explorer:index'...
gulp_1  | [18:05:49] Starting 'js:common:base'...
gulp_1  | [18:05:49] -> Rebundling hourglass_site/static_source/js/data-capture/index.js
gulp_1  | [18:05:49] { [SyntaxError: /calc/hourglass_site/static_source/js/data-capture/index.js: Unexpected token (6:15) while parsing file: /calc/hourglass_site/static_source/js/data-capture/index.js]
gulp_1  |   pos: 141,
gulp_1  |   loc: Position { line: 6, column: 15 },
gulp_1  |   _babel: true,
gulp_1  |   codeFrame: '  4 | // example: require(\'../common/upload\');\n  5 | \n> 6 | var boop = e() => { return "fooooo"; };\n    |                ^\n  7 | console.log(boop());\n  8 | ',
gulp_1  |   filename: '/calc/hourglass_site/static_source/js/data-capture/index.js',
gulp_1  |   stream: 
gulp_1  |    Labeled {
gulp_1  |      _readableState: 
gulp_1  |       ReadableState {
gulp_1  |         objectMode: true,
gulp_1  |         highWaterMark: 16,
gulp_1  |         buffer: [],
gulp_1  |         length: 0,
gulp_1  |         pipes: [Object],
gulp_1  |         pipesCount: 1,
gulp_1  |         flowing: true,
gulp_1  |         ended: false,
gulp_1  |         endEmitted: false,
gulp_1  |         reading: true,
gulp_1  |         sync: false,
gulp_1  |         needReadable: true,
gulp_1  |         emittedReadable: false,
gulp_1  |         readableListening: false,
gulp_1  |         resumeScheduled: false,
gulp_1  |         defaultEncoding: 'utf8',
gulp_1  |         ranOut: false,
gulp_1  |         awaitDrain: 0,
gulp_1  |         readingMore: false,
gulp_1  |         decoder: null,
gulp_1  |         encoding: null },
gulp_1  |      readable: true,
gulp_1  |      domain: null,
gulp_1  |      _events: 
gulp_1  |       { end: [Object],
gulp_1  |         error: [Object],
gulp_1  |         data: [Function: ondata],
gulp_1  |         _mutate: [Object] },
gulp_1  |      _eventsCount: 4,
gulp_1  |      _maxListeners: undefined,
gulp_1  |      _writableState: 
gulp_1  |       WritableState {
gulp_1  |         objectMode: true,
gulp_1  |         highWaterMark: 16,
gulp_1  |         needDrain: false,
gulp_1  |         ending: true,
gulp_1  |         ended: true,
gulp_1  |         finished: true,
gulp_1  |         decodeStrings: true,
gulp_1  |         defaultEncoding: 'utf8',
gulp_1  |         length: 0,
gulp_1  |         writing: false,
gulp_1  |         corked: 0,
gulp_1  |         sync: false,
gulp_1  |         bufferProcessing: false,
gulp_1  |         onwrite: [Function],
gulp_1  |         writecb: null,
gulp_1  |         writelen: 0,
gulp_1  |         bufferedRequest: null,
gulp_1  |         lastBufferedRequest: null,
gulp_1  |         pendingcb: 0,
gulp_1  |         prefinished: true,
gulp_1  |         errorEmitted: false,
gulp_1  |         bufferedRequestCount: 0,
gulp_1  |         corkedRequestsFree: [Object] },
gulp_1  |      writable: false,
gulp_1  |      allowHalfOpen: true,
gulp_1  |      _options: { objectMode: true },
gulp_1  |      _wrapOptions: { objectMode: true },
gulp_1  |      _streams: [ [Object] ],
gulp_1  |      length: 1,
gulp_1  |      label: 'deps' } }
gulp_1  | /calc/gulpfile.js:156
gulp_1  |         this.emit('end');
gulp_1  |             ^
gulp_1  | 
gulp_1  | TypeError: Cannot read property 'emit' of undefined
gulp_1  |     at Readable.<anonymous> (/calc/gulpfile.js:156:13)
gulp_1  |     at emitOne (events.js:95:20)
gulp_1  |     at Readable.emit (events.js:182:7)
gulp_1  |     at Labeled.<anonymous> (/node/node_modules/read-only-stream/index.js:28:44)
gulp_1  |     at emitOne (events.js:90:13)
gulp_1  |     at Labeled.emit (events.js:182:7)
gulp_1  |     at Labeled.<anonymous> (/node/node_modules/stream-splicer/index.js:130:18)
gulp_1  |     at emitOne (events.js:95:20)
gulp_1  |     at Labeled.emit (events.js:182:7)
gulp_1  |     at Deps.<anonymous> (/node/node_modules/stream-splicer/index.js:130:18)
calc_gulp_1 exited with code 1
```
</details>

Is it possible to have gulp not crash because of a syntax error, but rather to log the error and continue watching the files for changes?

**Comment by jseppi** [comment]
Sweet! I think there is a way to not exit on error, let me look into real quick.

**Comment by toolness** [comment]
Nice, it doesn't crash now!

Um except now I'm noticing that running `gulp build` exits with a `0` exit code even when I have a syntax error in my code... Mostly I'm concerned about Travis CI builds appearing to succeed even when there are syntax errors like this in our code, which hopefully won't happen since our test suites will be loading the JS bundles and failing because they can't be found... But if it's an easy fix to have gulp nonzero exit on `build` while continuously watching on `watch`, that would be nifty. If not it's ok too.

**Comment by jseppi** [comment]
Good point. It will now throw during `gulp build` or if it encounters an exception during the first time it runs through `gulp watch` (because the `js` task is called as a dependency before `isWatching` is set to `true`).

**Comment by toolness** [comment]
Hooray!!! I think this is good to go! :shipit:

**Pull request merged**